pallets/avn-anchor/src/lib.rs

```diff
@@ -6,62 +6,129 @@ extern crate alloc;
-    pub trait Config: frame_system::Config + pallet_avn::Config 
-    {
-        // type RuntimeEvent: From<Event<Self>>
-        //     + Into<<Self as frame_system::Config>::RuntimeEvent>
-        //     + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+    pub trait Config: frame_system::Config + pallet_avn::Config {
+        type RuntimeEvent: From<Event<Self>>
+            + Into<<Self as frame_system::Config>::RuntimeEvent>
+            + IsType<<Self as frame_system::Config>::RuntimeEvent>;
 
         type WeightInfo: WeightInfo;
     }
 
     #[pallet::pallet]
     pub struct Pallet<T>(_);
 
-    // #[pallet::event]
-    // #[pallet::generate_deposit(pub(super) fn deposit_event)]
-    // pub enum Event<T: Config> {}
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// A new chain handler was registered. [handler_account_id, chain_id, name]
+        ChainHandlerRegistered(T::AccountId, ChainId, BoundedVec<u8, ChainNameLimit>),
+        /// A chain handler was updated. [old_handler_account_id, new_handler_account_id, chain_id, name]
+        ChainHandlerUpdated(T::AccountId, T::AccountId, ChainId, BoundedVec<u8, ChainNameLimit>),
+    }
 
     #[pallet::error]
-    pub enum Error<T> {}
+    pub enum Error<T> {
+        ChainNotRegistered,
+        HandlerAlreadyRegistered,
+        UnauthorizedHandler,
+        NoAvailableChainId,
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn chain_handlers)]
+    pub type ChainHandlers<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, ChainData>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn next_chain_id)]
+    pub type NextChainId<T> = StorageValue<_, ChainId, ValueQuery>;
 
     #[pallet::call]
     impl<T: Config> Pallet<T> {
         #[pallet::weight(<T as pallet::Config>::WeightInfo::register_chain_handler())]
         #[pallet::call_index(0)]
         pub fn register_chain_handler(
             origin: OriginFor<T>,
-            // handler:
+            name: BoundedVec<u8, ChainNameLimit>,
         ) -> DispatchResult {
+            let handler = ensure_signed(origin)?;
+
+            ensure!(!ChainHandlers::<T>::contains_key(&handler), Error::<T>::HandlerAlreadyRegistered);
+
+            let chain_id = Self::get_next_chain_id()?;
+
+            let chain_data = ChainData {
+                chain_id,
+                name: name.clone(),
+            };
+
+            ChainHandlers::<T>::insert(&handler, chain_data);
+
+            Self::deposit_event(Event::ChainHandlerRegistered(handler, chain_id, name));
+
             Ok(())
         }
 
-        #[pallet::weight( <T as pallet::Config>::WeightInfo::update_chain_handler())]
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::update_chain_handler())]
         #[pallet::call_index(1)]
         pub fn update_chain_handler(
             origin: OriginFor<T>,
-            // handler:
+            new_handler: T::AccountId,
```

Comment:
Do you think that we should have a proof attached here with signed data of the new_handler?

Reply:
Not sure this is needed since the only actor that can set the new handler is the current handler